keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -589,6 +619,84 @@ associativeList.all(e, e.key1 == 'a' && e.key2 == 'b' && e.val == 100)
 associativeList.all(e, e.val == 100)
 ```
 
+### Access to Existing Object State
+
+#### Use Cases
+
+| Use Case                                                          | CEL Rule                                                                            | Examples |
+| --------                                                          | --------                                                                            | -------- |
+| Immutability                                                      | `self.foo == oldSelf.foo`                                                           ||
+| Prevent modification/removal once assigned                        | `oldSelf != 'bar' \|\| self == 'bar'` or `!has(oldSelf.field) \|\| has(self.field)` | [featureSet selection](https://github.com/openshift/kubernetes/blob/b426a3ffde5e7cd78483adeb5cb855f9f04dec5a/openshift-kube-apiserver/admission/customresourcevalidation/features/validate_features.go#L81-L88) |
+| Append-only set                                                   | `self.all(element, element in oldSelf)`                                             ||
+| If previous value was X, new value can only be A or B, not Y or Z | `oldSelf != 'X' \|\| self in ['A', 'B']`                                            | [Latency profile selection](https://github.com/openshift/enhancements/pull/977) |
+| Tolerating once-acceptable values ("ratcheting")                  | `self == oldSelf \|\| self in ["currently", "acceptable", "values"]`                ||
```

Comment:
I suggest removing this particular use-case.  If we are likely to add "run this rule against oldSelf and against newSelf" automatically to create ratcheting validation, we should not suggest a second way to do it.  I would even discourage this use-case for `oldSelf` and instead reserve `oldSelf` for those cases which are non-racheting.